### PR TITLE
Near Cache Eviction Implementation

### PR DESCRIPTION
**Pull request by serkan-ozal**
This PR contains eviction implementation for Near Cache

### CODE REVIEW

**Comment by serkan-ozal** [comment]
verify

**Comment by bilalyasar** [comment]
can you also update `ConfigXmlGenerator` class?

File: hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java [hunk 1]
```diff
@@ -130,18 +133,18 @@ public int getSize() {
         return size;
     }
 
-    public CacheEvictionConfig setSize(int size) {
+    public EvictionConfig setSize(int size) {
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java]: what about `size` smaller than 0? we can throw illegalArgumentException, or InvalidConfigurationException.

**Comment by bilalyasar** [comment]
can you write a test to `InvalidConfigurationTest`? we should test xsd changes.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/eviction/EvictionPolicyEvaluatorProvider.java [hunk 2]
```diff
@@ -45,11 +45,11 @@ private static void init() {
     /**
      * Gets the {@link EvictionPolicyEvaluator} implementation specified with <code>evictionPolicy</code>.
      *
```
Comment (metanet) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/eviction/EvictionPolicyEvaluatorProvider.java]: I think we can move getEvictionPolicyEvaluator method into EvictionPolicyType enum as a static method and remove this EvictionPolicyEvaluatorProvider class.

**Comment by serkan-ozal** [comment]
@metanet @bilalyasar thank you guys for your reviews. I will wait for some other comments if there is until tomorrow afternoon (if there will be no any other comment, I will accept PR is OK), then I will process your comments before merge.

File: hazelcast-client-new/src/main/java/com/hazelcast/client/config/XmlClientConfigBuilder.java [hunk 3]
```diff
@@ -167,15 +169,6 @@ public ClientConfig build(ClassLoader classLoader) {
         }
     }
 
-    ClientConfig build(ClientConfig clientConfig) {
```
Comment (bilalyasar) [inline hazelcast-client-new/src/main/java/com/hazelcast/client/config/XmlClientConfigBuilder.java]: why did you remove this method? actually, it is `package-private`, so not public. also we have similar method `build(Config config)` in server-side.

File: hazelcast-client-new/src/main/java/com/hazelcast/client/config/XmlClientConfigBuilder.java [hunk 4]
```diff
@@ -264,11 +256,36 @@ private void handleNearCache(Node node) {
...
                 NearCacheConfig.LocalUpdatePolicy policy = NearCacheConfig.LocalUpdatePolicy.valueOf(value);
                 nearCacheConfig.setLocalUpdatePolicy(policy);
+            } else if ("eviction".equals(nodeName)) {
+                nearCacheConfig.setEvictionConfig(getEvictionConfig(child));
             }
         }
         clientConfig.addNearCacheConfig(name, nearCacheConfig);
     }
 
+    private EvictionConfig getEvictionConfig(final org.w3c.dom.Node node) {
```
Comment (bilalyasar) [inline hazelcast-client-new/src/main/java/com/hazelcast/client/config/XmlClientConfigBuilder.java]: can you remove `org.w3c.dom` and import `org.w3c.dom` to header of class. only `Node` more readable.

File: hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java [hunk 5]
```diff
@@ -130,18 +133,18 @@ public int getSize() {
...
     }
 
-    public CacheMaxSizePolicy getMaxSizePolicy() {
+    public MaxSizePolicy getMaxSizePolicy() {
         return maxSizePolicy;
     }
 
-    public CacheEvictionConfig setMaxSizePolicy(CacheMaxSizePolicy maxSizePolicy) {
+    public EvictionConfig setMaxSizePolicy(MaxSizePolicy maxSizePolicy) {
         // Max-Size policy cannot be null
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java]: here in the case of `null` we must throw an exception.

Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java]: using hazelcast `ValidationUtil` static methods can be helpful here.

File: hazelcast-client-new/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java [hunk 6]
```diff
@@ -123,7 +123,31 @@ protected AbstractClientInternalCacheProxy(CacheConfig cacheConfig, ClientContex
         syncListenerRegistrations = new ConcurrentHashMap<CacheEntryListenerConfiguration, String>();
         syncLocks = new ConcurrentHashMap<Integer, CountDownLatch>();
 
-        NearCacheConfig nearCacheConfig = cacheConfig.getNearCacheConfig();
+        NearCacheConfig clientNearCacheConfig = clientContext.getClientConfig().getNearCacheConfig(name);
```
Comment (pveentjer) [inline hazelcast-client-new/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java]: This method is getting too long. Can you check if checkstyle is skipping this file?

File: hazelcast-client-new/src/test/java/com/hazelcast/client/cache/nearcache/ClientNearCacheTestSupport.java [hunk 7]
```diff
@@ -69,6 +71,13 @@ protected ClientConfig createClientConfig() {
         return new ClientConfig();
     }
 
+    protected NearCacheConfig createNearCacheConfig(InMemoryFormat inMemoryFormat) {
+        return
```
Comment (pveentjer) [inline hazelcast-client-new/src/test/java/com/hazelcast/client/cache/nearcache/ClientNearCacheTestSupport.java]: Please remove the linefeed after the return.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/NearCacheRecordMap.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package com.hazelcast.cache.impl.nearcache.impl;
+
+import com.hazelcast.cache.impl.eviction.EvictableStore;
+import com.hazelcast.cache.impl.nearcache.NearCacheRecord;
+
+import java.util.concurrent.ConcurrentMap;
+
+public interface NearCacheRecordMap<K, V extends NearCacheRecord>
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/NearCacheRecordMap.java]: Can you add some doc.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/SampleableNearCacheRecordMap.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache.impl.nearcache.impl;
+
+import com.hazelcast.cache.impl.eviction.impl.strategy.sampling.SampleableEvictableStore;
+import com.hazelcast.cache.impl.nearcache.NearCacheRecord;
+
+public interface SampleableNearCacheRecordMap<K, V extends NearCacheRecord>
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/SampleableNearCacheRecordMap.java]: Can you add some doc

File: hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/maxsize/EntryCountNearCacheMaxSizeChecker.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache.impl.nearcache.impl.maxsize;
+
+import com.hazelcast.cache.impl.maxsize.MaxSizeChecker;
+import com.hazelcast.cache.impl.nearcache.impl.NearCacheRecordMap;
+
+public class EntryCountNearCacheMaxSizeChecker implements MaxSizeChecker {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/maxsize/EntryCountNearCacheMaxSizeChecker.java]: Can you add some doc.

File: hazelcast-client-new/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java (hunk 6 above)
Comment (serkan-ozal) [inline hazelcast-client-new/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java]: ok

File: hazelcast-client-new/src/main/java/com/hazelcast/client/config/XmlClientConfigBuilder.java (hunk 3 above)
Comment (serkan-ozal) [inline hazelcast-client-new/src/main/java/com/hazelcast/client/config/XmlClientConfigBuilder.java]: it was not used so I removed it

File: hazelcast-client-new/src/main/java/com/hazelcast/client/config/XmlClientConfigBuilder.java (hunk 4 above)
Comment (serkan-ozal) [inline hazelcast-client-new/src/main/java/com/hazelcast/client/config/XmlClientConfigBuilder.java]: ok

File: hazelcast-client-new/src/test/java/com/hazelcast/client/cache/nearcache/ClientNearCacheTestSupport.java (hunk 7 above)
Comment (serkan-ozal) [inline hazelcast-client-new/src/test/java/com/hazelcast/client/cache/nearcache/ClientNearCacheTestSupport.java]: ok

File: hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/NearCacheRecordMap.java (hunk 8 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/NearCacheRecordMap.java]: sure

File: hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/SampleableNearCacheRecordMap.java (hunk 9 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/SampleableNearCacheRecordMap.java]: sure

File: hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/maxsize/EntryCountNearCacheMaxSizeChecker.java (hunk 10 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/nearcache/impl/maxsize/EntryCountNearCacheMaxSizeChecker.java]: sure

File: hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java (hunk 1 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java]: ok

File: hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java (hunk 5 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/config/EvictionConfig.java]: ok, I was not aware of `ValidationUtil`
